package/MDAnalysis/lib/formats/libdcd.pyx

```diff
@@ -0,0 +1,414 @@
+                py_remarks = py_remarks.decode('ascii', 'ignore')
+
+        py_remarks = "".join(s for s in py_remarks if s in string.printable)
+
+        return py_remarks
+
+    def _estimate_n_frames(self):
+        extrablocksize = 48 + 8 if self.charmm & DCD_HAS_EXTRA_BLOCK else 0
+        self.firstframesize = (self.n_atoms + 2) * self.n_dims * sizeof(float) + extrablocksize
+        self.framesize = ((self.n_atoms - self.nfixed + 2) * self.n_dims * sizeof(float) +
+                          extrablocksize)
+        filesize = path.getsize(self.fname)
+        # It's safe to use ftell, even though ftell returns a long, because the
+        # header size is < 4GB.
+        self.header_size = fio_ftell(self.fp)
+        # TODO: check that nframessize is larger then 0, the c-implementation
+        # used to do that.
+        nframessize = filesize - self.header_size - self.firstframesize
+        return nframessize / self.framesize + 1
+
+    @property
+    def periodic(self):
+          return bool((self.charmm & DCD_IS_CHARMM) and
+                      (self.charmm & DCD_HAS_EXTRA_BLOCK))
+
+
+    def read(self):
+        if self.reached_eof:
+            raise IOError('Reached last frame in DCD, seek to 0')
+        if not self.is_open:
+            raise IOError("No file open")
+        if self.mode != 'r':
+            raise IOError('File opened in mode: {}. Reading only allow '
+                               'in mode "r"'.format('self.mode'))
+
+        cdef np.ndarray xyz = np.empty((self.n_atoms, 3), dtype=FLOAT,
+                                       order='F')
+        cdef np.ndarray unitcell = np.empty(6, dtype=DOUBLE)
+
+        cdef FLOAT_T[::1] x = xyz[:, 0]
+        cdef FLOAT_T[::1] y = xyz[:, 1]
+        cdef FLOAT_T[::1] z = xyz[:, 2]
+
+        first_frame = self.current_frame == 0
+
+        cdef int lowerb = 0
+        cdef int upperb = self.n_atoms - 1
+
+        ok = read_dcdsubset(self.fp, self.n_atoms, lowerb, upperb,
+                          <FLOAT_T*> &x[0],
+                          <FLOAT_T*> &y[0], <FLOAT_T*> &z[0],
+                          <DOUBLE_T*> unitcell.data, self.nfixed, first_frame,
+                          self.freeind, self.fixedcoords,
+                          self.reverse_endian, self.charmm)
+        if ok != 0 and ok != -4:
+            raise IOError("Reading DCD header failed: {}".format(DCD_ERRORS[ok]))
+
+        # we couldn't read any more frames.
+        if ok == -4:
+            self.reached_eof = True
+            raise StopIteration
+
+        self.current_frame += 1
+
+        _ts_order = [0, 2, 5, 4, 3, 1]
+        uc = np.take(unitcell, _ts_order)
+        if np.any(uc < 0.) or np.any(uc[3:] > 180.):
+            # might be new CHARMM: box matrix vectors
+            H = unitcell
+            e1, e2, e3 = H[[0,1,3]],  H[[1,2,4]], H[[3,4,5]]
+            uc = triclinic_box(e1, e2, e3)
+
+        unitcell = uc
+
+        return DCDFrame(xyz, unitcell)
```

Comment:
Do we handle fixed coordinates in this function correctly? Are they injected into the xyz array by `read_dcdsubset`? Do we even have test files with fixed coordinates?